### PR TITLE
Correct the bug report for `cargo clippy --fix`

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

close https://github.com/rust-lang/cargo/issues/11877

Correct the bug report for `cargo clippy --fix`.

### How should we test and review this PR?
See the unit tests.

### CODE REVIEW

**Comment by weihanglo** [comment]
Some tests like `check_fixable_warning_for_clippy` do similar stuff I believe.

**Comment by hi-rustin** [comment]
@weihanglo Friendly ping~ 
Could you please take a look when you have time? Thanks!

File: src/cargo/core/compiler/job_queue/mod.rs [hunk 1]
```diff
@@ -630,7 +630,7 @@ impl<'cfg> DrainState<'cfg> {
                 self.bump_warning_count(id, emitted, fixable);
             }
             Message::FixDiagnostic(msg) => {
-                self.print.print(&msg)?;
+                self.print.print(&msg, &cx.bcx.rustc().workspace_wrapper)?;
```
Comment (weihanglo) [inline src/cargo/core/compiler/job_queue/mod.rs:633]: It is a bit weird to me we added a new argument `&Option<PathBuf>` only for accomplishing this fix. Other solutions might be either adding a doc comment for `fn print` to explain it, or passing `workspace_wrapper` upfront when instantiating the `DiagnosticPrinter`.

If you eventually go with the latter, I encourage you to add missing doc comments for `DiagnosticPrinter` and its fields :)

Not a blocker though.

File: tests/testsuite/fix.rs [hunk 2]
```diff
@@ -160,7 +159,157 @@ fn broken_fixes_backed_out() {
...
+                let mut args = std::env::args_os();
+                let _me = args.next().unwrap();
+                let rustc = args.next().unwrap();
+                let status = std::process::Command::new(rustc).args(args).status().unwrap();
+                std::process::exit(status.code().unwrap_or(1));
+            }
+            "#,
+        )
+        .build();
+    clippy_driver.cargo("build").run();
```
Comment (weihanglo) [inline tests/testsuite/fix.rs:199]: This part is largely a duplicate. Is it possible to share it a bit with `check_fixable_warning_for_clippy`?

File: tests/testsuite/fix.rs [hunk 3]
```diff
@@ -160,7 +159,157 @@ fn broken_fixes_backed_out() {
...
+                    // Also compatible for rustc invocations with `@path` argfile.
+                    let is_lib_rs = env::args_os()
+                        .map(PathBuf::from)
+                        .flat_map(|p| if let Some(p) = p.to_str().unwrap_or_default().strip_prefix("@") {
+                            fs::read_to_string(p).unwrap().lines().map(PathBuf::from).collect()
+                        } else {
+                            vec![p]
+                        })
+                        .any(|l| l == Path::new("src/lib.rs"));
+                    if is_lib_rs {
```
Comment (weihanglo) [inline tests/testsuite/fix.rs:244]: I don't want to repeat this as well. Could you find a way to share the construction of this `Project` with `broken_fixes_backed_out`?

**Review by weihanglo — COMMENTED** [review]
Thanks for the fix! I don't think we'll have more tools in a near future, so this hard-coded solution effectively fixes the issue IMO.

Only some minor suggestions.

**Comment by hi-rustin** [comment]
> Only some minor suggestions.

Thanks for your review!  💚 💙 💜 💛 ❤️ All addressed!

**Review by weihanglo — APPROVED** [review]
Looks pretty good. Thanks for the update!

**Pull request merged**